drake/core/Function.h

```diff
@@ -0,0 +1,148 @@
+
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include <iostream>
+#include <Eigen/Dense>
+#include "Gradient.h"
+
+namespace Drake {
+
+  /** InputOutputRelation
+   * \brief Tags which can be used to inform algorithms about underlying structure in a function
+   * e.g., linear, affine, polynomial, analytic, differentiable, continuous, measurable, and -- lastly -- arbitrary
+   *
+   * note: i considered using inheritance to capture the relationship, but passing around types at runtime
+   * was more of a pain than simply capturing the inheritance with the helper functions below.
+   */
+  struct InputOutputRelation {
+    enum Form {
+      ARBITRARY,
+      DIFFERENTIABLE,
+      POLYNOMIAL,
+      AFFINE,
+      LINEAR,
+      CONSTANT,
+      ZERO
+      // todo: add more (rational, ...)
+    };
+    Form form;
+    // todo: add sparsity info
+
+    InputOutputRelation(Form f) : form(f) { };
+
+    static bool isa(const Form &f, const Form &base) {
+      if (f == base || base == ARBITRARY) return true;
+      if (f == ARBITRARY) return false;
+      return isa(derivesFrom(f), base);
+    }
+
+    bool isa(Form base) { return isa(form, base); }
+
+    static Form leastCommonAncestor(const Form &f1, const Form &f2) {
+      if (f1 == ARBITRARY || f2 == ARBITRARY) return ARBITRARY;
+      if (isa(f2, f1)) return f1;
+      return leastCommonAncestor(derivesFrom(f1), f2);
+    }
+
+    static Form leastCommonAncestor(std::initializer_list<Form> forms) {
+      if (forms.size() < 1) throw std::runtime_error("leastCommonAncestor requires at least one argument");
+      if (forms.size() == 1) return *forms.begin();
+
+      std::initializer_list<Form>::const_iterator iter = forms.begin();
+      Form ret = *iter;
+      while (++iter != forms.end()) {
+        ret = leastCommonAncestor(ret, *iter);
+      }
+      return ret;
+    }
+
+    static InputOutputRelation composeWith(const InputOutputRelation &g,
+                                           const InputOutputRelation &f) { // composition of functions y = g(f(x))
+      return InputOutputRelation(leastCommonAncestor(g.form, f.form));
```

Comment:
Don't know if we're going to care about this case, but imagine composing an arbitrary function and a zero function. The best type of the result would be a zero function, not an arbitrary function.